aw_watcher_window/lib.py

```diff
@@ -1,19 +1,77 @@
 import sys
+import json
+import logging
 from typing import Optional
 
+logger = logging.getLogger(__name__)
 
-def get_current_window_linux() -> Optional[dict]:
-    from . import xlib
-    window = xlib.get_current_window()
 
-    if window is None:
-        cls = "unknown"
-        name = "unknown"
-    else:
-        cls = xlib.get_window_class(window)
-        name = xlib.get_window_name(window)
+class Linux:
+    def __init__(self):
+        try:
+            import pydbus
+            self.bus = pydbus.SessionBus()
+        except ModuleNotFoundError:
+            logger.info("pydbus not installed, GNOME-Shell Wayland support disabled")
+            self.bus = False
+            self.gnome_shell = None
+        
+        if self.bus:
+            import gi.repository.GLib
+            try:
+                self.gnome_shell = self.bus.get("org.gnome.Shell")
+                self._setup_gnome()
+            except gi.repository.GLib.Error:
+                self.gnome_shell = None
+
+    def get_current_window(self) -> dict:
+        if self.gnome_shell:
+            return self.get_current_window_gnome_shell()
+        
+        return self.get_current_window_x11()
+    
+    def _setup_gnome(self) -> None:
+        js_code = """
+        global._aw_current_window = () => {
```

Comment:
Nice!
Much better solution now, no race conditions so seems like it should be reliable.